catlas/nuclearity.py

```diff
@@ -0,0 +1,122 @@
+from graph_tool import topology
+import numpy as np
+
+
+def get_nuclearity(entry):
+    """
+    Function to get the nuclearity for each element in a surface
+    Args:
+        entry: a catlas-like entry object
+
+    Returns:
+        entry: a catlas-like entry object with nuclearity info added
+    """
+    elements = entry["bulk_elements"]
+    slab_atoms = entry["slab_surface_object"].surface_atoms
+    replicated_slab_atoms = slab_atoms.repeat((2, 2, 1))
+
+    # Grab connectivity matricies
+    overall_connectivity_matrix = get_connectivity_matrix(slab_atoms)
+    overall_connectivity_matrix_rep = get_connectivity_matrix(replicated_slab_atoms)
+
+    # Grab surface atom idxs
+    surface_indices = [idx for idx, tag in enumerate(slab_atoms.get_tags()) if tag == 1]
+    surface_indices_rep = [
+        idx for idx, tag in enumerate(replicated_slab_atoms.get_tags()) if tag == 1
+    ]
+
+    # Iterate over atoms and assess nuclearity
+    output_dict = {}
+    for element in elements:
+        connected_surface_atoms = [
+            atom.symbol == element and atom.index in surface_indices
+            for atom in slab_atoms
+        ]
+        connected_surface_atoms_rep = [
+            atom.symbol == element and atom.index in surface_indices_rep
+            for atom in replicated_slab_atoms
+        ]
+
+        if sum(connected_surface_atoms) == 0:
+            output_dict[element] = {"nuclearity": 0, "nuclearities": []}
+
+        else:
+            hist = get_nuclearity_neighbor_counts(
+                connected_surface_atoms, overall_connectivity_matrix
+            )
+            hist_rep = get_nuclearity_neighbor_counts(
+                connected_surface_atoms_rep, overall_connectivity_matrix_rep
+            )
+            output_dict[element] = evaluate_infiniteness(hist, hist_rep)
+    entry["nuclearity_info"] = output_dict
+    return entry
+
+
+def get_nuclearity_neighbor_counts(connected_surface_atoms, connectivity_matrix):
+    """
+    Function that counts the like surface neighbors for surface atoms
+    Args:
+        connected_surface_atoms: list of all surface atoms which are of a specific element
+        connectivity_matrix: matrix memorializing which atoms in the slab are connected
+
+    Returns:
+        hist: counts of neighbor groups
+    """
+    connectivity_matrix = connectivity_matrix[connected_surface_atoms, :]
+    connectivity_matrix = connectivity_matrix[:, connected_surface_atoms]
+    graph = gt.Graph(directed=False)
+    graph.add_vertex(n=connectivity_matrix.shape[0])
+    graph.add_edge_list(np.transpose(connectivity_matrix.nonzero()))
+    labels, hist = topology.label_components(graph, directed=False)
+    return hist
+
+
+def evaluate_infiniteness(hist, hist_rep):
+    """
+    Function that compares the connected counts between the minimal slab
+    and a repeated slab to classify the type of infiniteness.
+    Args:
+        hist: counts of neighbor types in the minimal slab
```

Comment:
Is this a list of lists of ints?

Reply:
its a list of ints. I will make that more obvious. Thats why I take the max() of it